src/threading/lambda.h

```diff
@@ -0,0 +1,64 @@
+// Minetest
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#pragma once
+
+#include <exception>
+#include <functional>
+#include <memory>
+#include "debug.h"
+#include "threading/thread.h"
+
+/**
+ * Class returned by `runInThread`.
+ *
+ * Provides the usual thread methods along with `rethrow()`.
+*/
+class LambdaThread : public Thread
+{
+	friend std::unique_ptr<LambdaThread> runInThread(
+		const std::function<void()> &, const std::string &);
+public:
+	/// Re-throw a caught exception, if any. Can only be called after thread exit.
+	void rethrow()
+	{
+		sanity_check(!isRunning());
+		if (m_exptr)
+			std::rethrow_exception(m_exptr);
+	}
+
+private:
+	// hide methods
+	using Thread::Thread;
+	using Thread::start;
+
+	std::function<void()> m_fn;
+	std::exception_ptr m_exptr;
+
+	void *run()
+	{
+		try {
+			m_fn();
+		} catch(...) {
+			m_exptr = std::current_exception();
+		}
+		return nullptr;
+	};
+};
+
+/**
+ * Run a lambda in a separate thread.
+ *
+ * Exceptions will be caught.
+ * @param fn function to run
+ * @param thread_name name for thread
+ * @return thread object of type `LambdaThread`
+*/
+std::unique_ptr<LambdaThread> runInThread(const std::function<void()> &fn,
+	const std::string &thread_name = "")
+{
+	auto t = std::make_unique<LambdaThread>(thread_name);
```

Comment:
I had this before but it caused an error because the constructor is private.

I'm in fact wondering why it's working now.

Reply:
It looks like the constructor isn't actually private? Adding these two lines to `Game::shutdown()` doesn't result in any errors:

```c++
LambdaThread test1("test");
LambdaThread *test2 = new LambdaThread("test");
```

Reply:
Looks like `using Thread::Thread` doesn't work to hide the constructor then.

could change that back to `LambdaThread(const std::string &name) : Thread(name) {}` or we could opt to not care